File: platforms/windows/Dimmy.Windows/Helpers/WavPeaks.cs
using System;
using System.IO;
using System.Text;
using System.Text.Json;

namespace Dimmy.Windows.Helpers;

/// Minimal WAV reader that produces normalized peak buckets for
/// waveform rendering. Only handles the formats Dimmy itself writes
/// (16 kHz mono PCM_S16LE, 16 kHz mono PCM_F32LE) plus stereo PCM_S16LE
/// for arbitrary user-loaded files. Anything else returns an empty
/// array — caller falls back to "no waveform" UI.
///
/// Output: float[bucketCount] of values in [0..1] representing the
/// peak absolute amplitude of each bucket. The History detail panel
/// renders this as filled rectangles in a Canvas.
public static class WavPeaks
{
    /// Read peaks for ANY audio file the loader supports. WAV stays on
    /// the in-process hand-rolled fast path (no FFI hop, no Rust
    /// decoder spin-up cost); everything else (m4a / mp3 / aac / flac /
    /// ogg) goes through the Rust `dimmy_compute_audio_peaks` FFI
    /// which shares the Symphonia decoder used by `dimmy_transcribe_file`.
    ///
    /// Returns the same shape `ReadPeaks` always has (float[bucketCount]
    /// in [0..1]), or an empty array if both paths fail. The caller
    /// (History detail panel) treats "empty" as "no waveform" UI.
    public static float[] ReadPeaksAny(string path, int bucketCount)
    {
        if (string.IsNullOrEmpty(path) || bucketCount <= 0)
            return Array.Empty<float>();
        var ext = Path.GetExtension(path);
        if (!string.IsNullOrEmpty(ext)
            && ext.Equals(".wav", StringComparison.OrdinalIgnoreCase))
        {
            var wavPeaks = ReadPeaks(path, bucketCount);
            if (wavPeaks.Length > 0) return wavPeaks;
            // Fall through if the WAV magic was missing — symphonia
            // can recover from some odd containers (mp4 mis-named .wav)
            // where hound would just fail silently.
        }
        return ReadPeaksViaFfi(path, bucketCount);
    }

    /// Pure-FFI path used for non-WAV formats (m4a / mp3 / aac / flac /
    /// ogg). Allocates a JSON buffer sized for `bucketCount` 32-bit
    /// floats + framing overhead (~16 bytes per number including
    /// commas and brackets), with a small extra for the duration_secs
    /// field. Empty array on any error — host falls back to "no
    /// waveform".
    private static float[] ReadPeaksViaFfi(string path, int bucketCount)
    {
        try
        {
            // Heuristic: each peak serialises to at most ~14 chars
            // ("0.1234567,"), plus 64 bytes for the duration + brackets
            // + keys + null terminator. Round up to the next 4 KB
            // boundary for safety on very wide canvases.
            int needed = bucketCount * 16 + 128;
            int bufLen = ((needed + 4095) / 4096) * 4096;
            var buf = new byte[bufLen];
            int rc = Interop.DimmyNative.dimmy_compute_audio_peaks(
                path, bucketCount, buf, bufLen);
            if (rc <= 0) return Array.Empty<float>();
            string json = Encoding.UTF8.GetString(buf, 0, rc);
            using var doc = JsonDocument.Parse(json);
            if (!doc.RootElement.TryGetProperty("peaks", out var arr)
                || arr.ValueKind != JsonValueKind.Array)
                return Array.Empty<float>();
            int n = arr.GetArrayLength();
            var peaks = new float[n];
            int i = 0;
            foreach (var el in arr.EnumerateArray())
            {
                peaks[i++] = (float)el.GetDouble();
            }
            return peaks;
        }
        catch
        {
            return Array.Empty<float>();
        }
    }

    public static float[] ReadPeaks(string path, int bucketCount)
    {
        try
        {
            if (string.IsNullOrEmpty(path) || !File.Exists(path))
                return Array.Empty<float>();
            if (bucketCount <= 0) return Array.Empty<float>();

            using var fs = File.OpenRead(path);
            using var br = new BinaryReader(fs);

            if (new string(br.ReadChars(4)) != "RIFF") return Array.Empty<float>();
            br.ReadUInt32(); // file size
            if (new string(br.ReadChars(4)) != "WAVE") return Array.Empty<float>();

            ushort fmtTag = 0;
            ushort channels = 1;
            uint sampleRate = 0;
            ushort bitsPerSample = 0;
            long dataOffset = -1;
            uint dataSize = 0;

            // Walk chunks until we see fmt + data.
            while (fs.Position + 8 <= fs.Length)
            {
                var id = new string(br.ReadChars(4));
                var size = br.ReadUInt32();
                if (id == "fmt ")
                {
                    fmtTag = br.ReadUInt16();
                    channels = br.ReadUInt16();
                    sampleRate = br.ReadUInt32();
                    br.ReadUInt32(); // byte rate
                    br.ReadUInt16(); // block align
                    bitsPerSample = br.ReadUInt16();
                    var consumed = 16u;
                    if (size > consumed) fs.Seek(size - consumed, SeekOrigin.Current);
                }
                else if (id == "data")
                {
                    dataOffset = fs.Position;
                    dataSize = size;
                    break;
                }
                else
                {
                    fs.Seek(size, SeekOrigin.Current);
                }
            }

            if (dataOffset < 0 || channels == 0 || bitsPerSample == 0)
                return Array.Empty<float>();

            int bytesPerSample = bitsPerSample / 8;
            int frameSize = bytesPerSample * channels;
            if (frameSize == 0) return Array.Empty<float>();
            long totalFrames = dataSize / (uint)frameSize;
            if (totalFrames <= 0) return Array.Empty<float>();

            var peaks = new float[bucketCount];
            long framesPerBucket = Math.Max(1, totalFrames / bucketCount);

            fs.Seek(dataOffset, SeekOrigin.Begin);
            for (int b = 0; b < bucketCount; b++)
            {
                float peak = 0f;
                long startFrame = b * framesPerBucket;
                long endFrame = Math.Min(totalFrames, (b + 1) * framesPerBucket);
                long framesToRead = endFrame - startFrame;
                if (framesToRead <= 0) break;

                fs.Seek(dataOffset + startFrame * frameSize, SeekOrigin.Begin);
                for (long f = 0; f < framesToRead; f++)
                {
                    float maxChannel = 0f;
                    for (int c = 0; c < channels; c++)
                    {
                        float s = ReadSample(br, fmtTag, bitsPerSample);
                        var a = Math.Abs(s);
                        if (a > maxChannel) maxChannel = a;
                    }
                    if (maxChannel > peak) peak = maxChannel;
                }
                peaks[b] = peak > 1f ? 1f : peak;
            }
            return peaks;
        }
        catch
        {
            return Array.Empty<float>();
        }
    }

    /// <summary>
    /// Return the audio duration in seconds, or 0 if the file is
    /// missing / unreadable / not a WAV. Walks the same header as
    /// <see cref="ReadPeaks"/> but stops after the format/data
    /// chunk metadata, so it's cheap (no per-sample work).
    /// </summary>
    public static double ReadDurationSecs(string path)
    {
        try
        {
            if (string.IsNullOrEmpty(path) || !File.Exists(path)) return 0;
            using var fs = File.OpenRead(path);
            using var br = new BinaryReader(fs);
            if (new string(br.ReadChars(4)) != "RIFF") return 0;
            br.ReadUInt32();
            if (new string(br.ReadChars(4)) != "WAVE") return 0;

            ushort channels = 1;
            uint sampleRate = 0;
            ushort bitsPerSample = 0;
            uint dataSize = 0;

            while (fs.Position + 8 <= fs.Length)
            {
                var id = new string(br.ReadChars(4));
                var size = br.ReadUInt32();
                if (id == "fmt ")
                {
                    br.ReadUInt16(); // tag
                    channels = br.ReadUInt16();
                    sampleRate = br.ReadUInt32();
                    br.ReadUInt32(); // byte rate
                    br.ReadUInt16(); // block align
                    bitsPerSample = br.ReadUInt16();
                    var consumed = 16u;
                    if (size > consumed) fs.Seek(size - consumed, SeekOrigin.Current);
                }
                else if (id == "data")
                {
                    dataSize = size;
                    break;
                }
                else
                {
                    fs.Seek(size, SeekOrigin.Current);
                }
            }

            if (sampleRate == 0 || bitsPerSample == 0 || channels == 0) return 0;
            int bytesPerSample = bitsPerSample / 8;
            int frameSize = bytesPerSample * channels;
            if (frameSize == 0) return 0;
            long totalFrames = dataSize / (uint)frameSize;
            return (double)totalFrames / sampleRate;
        }
        catch { return 0; }
    }

    private static float ReadSample(BinaryReader br, ushort fmtTag, ushort bits)
    {
        // 1=PCM int, 3=IEEE float
        if (fmtTag == 3 && bits == 32)
            return br.ReadSingle();
        if (bits == 16)
            return br.ReadInt16() / 32768f;
        if (bits == 8)
            return (br.ReadByte() - 128) / 128f;
        if (bits == 24)
        {
            int b0 = br.ReadByte();
            int b1 = br.ReadByte();
            int b2 = (sbyte)br.ReadByte();
            int v = (b2 << 16) | (b1 << 8) | b0;
            return v / 8388608f;
        }
        if (bits == 32 && fmtTag == 1)
            return br.ReadInt32() / 2147483648f;
        // Unknown — skip frame to avoid an infinite loop, caller treats peak as 0.
        br.ReadBytes(bits / 8);
        return 0f;
    }
}

File: platforms/macos/Dimmy/Views/Meeting/WavPeaks.swift
import Foundation

// MARK: - WavPeaks
//
// Mac port of Win `Helpers/WavPeaks.cs`. Reads a WAV file, walks the
// `fmt ` + `data` chunks, decodes samples (PCM int8/16/24/32 + IEEE
// float32), and bucketises the maximum absolute amplitude into
// `bucketCount` normalized peaks in [0, 1]. Used by the meeting Done
// view to draw a click-to-seek waveform.
//
// Hermetic: no FFI, no dispatch, no allocations beyond the output
// buffer + a 16 KiB read buffer for the data chunk. Returns an empty
// array on any parse failure — the UI falls back to a thin slider.

enum WavPeaks {
    /// Read `path` and produce `bucketCount` normalized peaks. Returns
    /// `[]` on any error (missing file, unknown format, truncated WAV).
    static func readPeaks(path: String, bucketCount: Int) -> [Float] {
        guard bucketCount > 0,
              FileManager.default.fileExists(atPath: path),
              let data = try? Data(contentsOf: URL(fileURLWithPath: path),
                                    options: [.alwaysMapped])
        else { return [] }
        return parse(data: data, bucketCount: bucketCount)
    }

    static func parse(data: Data, bucketCount: Int) -> [Float] {
        // Need at least the RIFF header + fmt + data headers.
        guard data.count >= 44 else { return [] }
        guard read4cc(data, at: 0) == "RIFF",
              read4cc(data, at: 8) == "WAVE"
        else { return [] }

        var fmtTag: UInt16 = 0
        var channels: UInt16 = 1
        var bitsPerSample: UInt16 = 0
        var dataOffset: Int = -1
        var dataSize: Int = 0

        var pos = 12
        while pos + 8 <= data.count {
            let id = read4cc(data, at: pos)
            let size = Int(readU32LE(data, at: pos + 4))
            let bodyStart = pos + 8
            switch id {
            case "fmt ":
                guard bodyStart + 16 <= data.count else { return [] }
                fmtTag = readU16LE(data, at: bodyStart)
                channels = readU16LE(data, at: bodyStart + 2)
                // sampleRate at +4, byteRate at +8, blockAlign at +12
                bitsPerSample = readU16LE(data, at: bodyStart + 14)
            case "data":
                dataOffset = bodyStart
                dataSize = size
            default:
                break
            }
            if dataOffset >= 0 { break }
            pos = bodyStart + size
            // WAV chunk bodies are padded to even length.
            if size % 2 == 1 { pos += 1 }
        }

        guard dataOffset >= 0,
              channels > 0,
              bitsPerSample > 0,
              bitsPerSample % 8 == 0
        else { return [] }

        let bytesPerSample = Int(bitsPerSample / 8)
        let frameSize = bytesPerSample * Int(channels)
        guard frameSize > 0 else { return [] }
        let totalFrames = min(dataSize, data.count - dataOffset) / frameSize
        guard totalFrames > 0 else { return [] }

        var peaks = [Float](repeating: 0, count: bucketCount)
        let framesPerBucket = max(1, totalFrames / bucketCount)

        for b in 0..<bucketCount {
            let startFrame = b * framesPerBucket
            let endFrame = min(totalFrames, (b + 1) * framesPerBucket)
            if startFrame >= endFrame { break }
            var peak: Float = 0
            for f in startFrame..<endFrame {
                let frameOffset = dataOffset + f * frameSize
                var maxCh: Float = 0
                for c in 0..<Int(channels) {
                    let s = readSample(
                        data: data,
                        offset: frameOffset + c * bytesPerSample,
                        fmtTag: fmtTag,
                        bits: Int(bitsPerSample)
                    )
                    let a = abs(s)
                    if a > maxCh { maxCh = a }
                }
                if maxCh > peak { peak = maxCh }
            }
            peaks[b] = min(peak, 1)
        }
        return peaks
    }

    // MARK: - Decoders

    private static func readSample(data: Data, offset: Int, fmtTag: UInt16, bits: Int) -> Float {
        // 1=PCM int, 3=IEEE float (per WAVE_FORMAT_PCM / IEEE_FLOAT).
        if fmtTag == 3, bits == 32 {
            return readF32LE(data, at: offset)
        }
        if bits == 16 {
            let v = readI16LE(data, at: offset)
            return Float(v) / 32_768.0
        }
        if bits == 8 {
            // 8-bit WAV is unsigned with center 128.
            return (Float(data[offset]) - 128.0) / 128.0
        }
        if bits == 24 {
            let b0 = Int32(data[offset])
            let b1 = Int32(data[offset + 1])
            let b2 = Int32(Int8(bitPattern: data[offset + 2]))  // sign-extend
            let v = (b2 << 16) | (b1 << 8) | b0
            return Float(v) / 8_388_608.0
        }
        if bits == 32, fmtTag == 1 {
            let v = readI32LE(data, at: offset)
            return Float(v) / 2_147_483_648.0
        }
        return 0
    }

    // MARK: - Little-endian readers

    private static func read4cc(_ data: Data, at offset: Int) -> String {
        guard offset + 4 <= data.count else { return "" }
        let bytes = data[offset..<(offset + 4)]
        return String(bytes: bytes, encoding: .ascii) ?? ""
    }

    private static func readU16LE(_ data: Data, at offset: Int) -> UInt16 {
        UInt16(data[offset]) | (UInt16(data[offset + 1]) << 8)
    }

    private static func readU32LE(_ data: Data, at offset: Int) -> UInt32 {
        UInt32(data[offset])
            | (UInt32(data[offset + 1]) << 8)
            | (UInt32(data[offset + 2]) << 16)
            | (UInt32(data[offset + 3]) << 24)
    }

    private static func readI16LE(_ data: Data, at offset: Int) -> Int16 {
        Int16(bitPattern: readU16LE(data, at: offset))
    }

    private static func readI32LE(_ data: Data, at offset: Int) -> Int32 {
        Int32(bitPattern: readU32LE(data, at: offset))
    }

    private static func readF32LE(_ data: Data, at offset: Int) -> Float {
        Float(bitPattern: readU32LE(data, at: offset))
    }
}
